syntax = "proto3";

option java_multiple_files = true;
option java_package = "com.vegvisir.core.datatype.proto";
option java_outer_classname = "VegvisirCoreDatatypeProto";
option objc_class_prefix = "VEGVISIR";

package vegvisir.protos;

import "vegvisir/protos/common.proto";
import "vegvisir/protos/vegvisirlower.proto";

message Certificate {
    string userid = 9;
    bytes public_key = 10;
    bytes private_key = 11;
    bytes signature = 12;
}

message Block {
    message GenesisBlock{
        string userid = 1;
        float timestamp = 2;
        Certificate ca_certificate = 3;
        repeated Certificate certlist = 4;
        // Signed by CA, users not longer get access to CA certificate.
        bytes signature = 5;
    }

    message Transaction {
        string userid = 1;
        double timestamp = 2;
        int32 recordid = 3;
        bytes comment = 4;
    }
  
    message UserBlock {
        string userid = 1;
        Timestamp timestamp = 2;
        Location location = 3;
        repeated charlotte.Reference parents = 4;
        repeated Transaction transactions = 5;
    }
    
    charlotte.Signature signature = 1;

    oneof blocktype_oneof {
        GenesisBlock genesis_block = 2;
        UserBlock user_block = 3;
    }

}

message Blockchain {
    message UseridAssociations {
        string userid = 5;
    }

    message Keystore {
        Certificate ca_cert = 6;
        repeated UseridAssociations userids = 7;
        repeated Certificate usercerts = 8;
    }

    Block.GenesisBlock genesis = 9;
    Keystore keystore = 10;
    // Frontier nodes can be built from the genesis block
    // Crdt can be built from scratch at the beginning.
}


